pkg/providers/vsphere/setupuser/validator.go

```diff
@@ -0,0 +1,186 @@
+)
+
+type connection struct {
+	Server   string `yaml:"server"`
+	Insecure bool   `yaml:"insecure"`
+}
+
+type objects struct {
+	Networks      []string `yaml:"networks"`
+	Datastores    []string `yaml:"datastores"`
+	ResourcePools []string `yaml:"resourcePools"`
+	Folders       []string `yaml:"folders"`
+	Templates     []string `yaml:"templates"`
+}
+
+type vSphereUserSpec struct {
+	Datacenter    string     `yaml:"datacenter"`
+	VSphereDomain string     `yaml:"vSphereDomain"`
+	Connection    connection `yaml:"connection"`
+	Objects       objects    `yaml:"objects"`
+	// Below are optional fields with defaults
+	Username   string `yaml:"username"`
+	GroupName  string `yaml:"group"`
+	GlobalRole string `yaml:"globalRole"`
+	UserRole   string `yaml:"userRole"`
+	AdminRole  string `yaml:"adminRole"`
+}
+
+type VSphereSetupUserConfig struct {
+	ApiVersion string          `yaml:"apiVersion"`
+	Kind       string          `yaml:"kind"`
+	Spec       vSphereUserSpec `yaml:"spec"`
+}
+
+func GenerateConfig(ctx context.Context, filepath string) (*VSphereSetupUserConfig, error) {
+	c, err := readConfig(ctx, filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validate(ctx, c)
+	if err != nil {
+		return nil, err
+	}
+
+	setDefaults(ctx, c)
+
+	return c, nil
+}
+
+func readConfig(ctx context.Context, filepath string) (*VSphereSetupUserConfig, error) {
+	file, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read file %s, err = %v", filepath, err)
+	}
+
+	c := VSphereSetupUserConfig{}
+
+	if err = yaml.Unmarshal(file, &c); err != nil {
+		return nil, fmt.Errorf("Failed to parse %s, err = %v", filepath, err)
+	}
+
+	return &c, nil
+}
+
+func validate(ctx context.Context, c *VSphereSetupUserConfig) error {
+	results := []validations.ValidationResult{
+		{
+			Name:        "validate datacenter",
+			Remediation: "",
+			Err:         validateDatacenter(ctx, c),
+		},
+		{
+			Name:        "validate vspheredomain",
+			Remediation: "",
+			Err:         validateVSphereDomain(ctx, c),
+		},
+		{
+			Name:        "validate connection",
+			Remediation: "",
+			Err:         validateConnection(ctx, c),
+		},
+	}
+
+	errs := []string{}
+	for _, r := range results {
+		if r.Err != nil {
+			errs = append(errs, r.Err.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return &validations.ValidationError{Errs: errs}
+	}
+
+	return nil
+}
+
+func validateDatacenter(ctx context.Context, c *VSphereSetupUserConfig) error {
+	if c.Spec.Datacenter == "" {
+		return fmt.Errorf("datacenter cannot be empty")
+	}
+	return nil
+}
+
+func validateVSphereDomain(ctx context.Context, c *VSphereSetupUserConfig) error {
+	if c.Spec.VSphereDomain == "" {
+		return fmt.Errorf("vSphereDomain cannot be empty")
+	}
+	return nil
+}
+
+func validateConnection(ctx context.Context, c *VSphereSetupUserConfig) error {
+	if c.Spec.Connection.Server == "" {
+		return fmt.Errorf("server cannot be empty")
+	}
+	return nil
+}
+
```

Comment:
My only thought for doing it this way is that it would be nice to have and use a ~standard framework for validating inputs in our CLI.

I was riffing off of [this PR ](https://github.com/aws/eks-anywhere/pull/3119/files) and trying to keep things similar.

Reply:
Do you mind if we stick to the framework? What do you think?

Reply:
A `validate` function could still be used in some framework - that's really about signatures etc. In any case, this isn't being used with a framework as is so I still feel like it would be simpler and clearer to have a `validate(cfg *VSphereSetupUserConfig) error`.

Re 'standard frameworks' we don't really have one that serves our needs. There have been various attempts but we've never got it quite right.

Reply:
Updated 👍